flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/sink/dynamic/DynamicCommitter.java

```diff
@@ -329,6 +352,52 @@ private void commitDeltaTxn(
     }
   }
 
+  static class MaxCommittedCheckpointIdValidator implements SnapshotAncestryValidator {
+    private final long stagedCheckpointId;
+    private final String flinkJobId;
+    private final String flinkOperatorId;
+
+    private String errorMessage = null;
+
+    MaxCommittedCheckpointIdValidator(
+        long stagedCheckpointId, String flinkJobId, String flinkOperatorId) {
+      this.stagedCheckpointId = stagedCheckpointId;
+      this.flinkJobId = flinkJobId;
+      this.flinkOperatorId = flinkOperatorId;
+    }
+
+    @Override
+    public Boolean apply(Iterable<Snapshot> baseSnapshots) {
+      FlinkSnapshot maxCommittedSnapshot =
+          getMaxCommittedSnapshot(baseSnapshots, flinkJobId, flinkOperatorId);
+      if (maxCommittedSnapshot != null
+          && maxCommittedSnapshot.checkpointId() == stagedCheckpointId) {
+        errorMessage =
+            String.format(
+                "The new parent snapshot '%s' has duplicate '%s': '%s'."
+                    + "\nThis can happen, for example, when using the REST catalog: if the previous commit request failed"
+                    + " in the Flink client but succeeded on the server after the Flink job decided to retry it with the new request."
+                    + "\nFlink should retry this exception, and the committer should skip the duplicate request during the next retry.",
+                maxCommittedSnapshot.snapshot().snapshotId(),
+                MAX_COMMITTED_CHECKPOINT_ID,
+                maxCommittedSnapshot.checkpointId());
```

Comment:
English is not my native language, but this error message seems a bit hard to understand.

What about something like this:
```
The %s branch of the %s table already contains changes for %s checkpoint.
This can occur when a failure prevents the committer from receiving confirmation of a successful commit, causing the Flink job to retry committing the same set of changes. After a job restart, the issue should be resolved, and only new changes will be committed to the table.
```

This would also allow us to not carry the `FlinkSnapshot.snapshot()` around. We could revert to store only the `lastCommittedCheckpointId`